### ISSUE TITLE
failure to fold comparisons after llvm.assume / __builtin_assume with unsigned value

### ISSUE THREAD

**Issue by zmodem** [llvm/llvm-project#62149]
Consider:

```
void g();

void f(int *p) {
  __builtin_assume(*p > 100);
  if (--*p < 42) g();
  if (--*p < 42) g();
  if (--*p < 42) g();
  if (--*p < 42) g();
  if (--*p < 42) g();
}
```

LLVM will fold all the comparisons nicely:

```
$ clang -O2 -S -o - a.c
[...]
	addl	$-5, (%rdi)
	retq
```

But if the value is unsigned, it will only fold the first three:

```
$ clang -O2 -S -o - a.c -Dint=unsigned
	.text
	.file	"a.c"
	.globl	f                               # -- Begin function f
	.p2align	4, 0x90
	.type	f,@function
f:                                      # @f
	.cfi_startproc
# %bb.0:                                # %entry
	pushq	%rbx
	.cfi_def_cfa_offset 16
	.cfi_offset %rbx, -16
	movq	%rdi, %rbx
	movl	(%rdi), %eax
	addl	$-4, %eax
	movl	%eax, (%rdi)
	cmpl	$41, %eax
	ja	.LBB0_2
# %bb.1:                                # %if.then12
	xorl	%eax, %eax
	callq	g@PLT
	movl	(%rbx), %eax
.LBB0_2:                                # %if.end13
	decl	%eax
	movl	%eax, (%rbx)
	cmpl	$41, %eax
	ja	.LBB0_3
# %bb.4:                                # %if.then16
	xorl	%eax, %eax
	popq	%rbx
	.cfi_def_cfa_offset 8
	jmp	g@PLT                           # TAILCALL
.LBB0_3:                                # %if.end17
	.cfi_def_cfa_offset 16
	popq	%rbx
	.cfi_def_cfa_offset 8
	retq
```

(GCC folds all of them in both cases, see https://godbolt.org/z/ssPEfM5Gd)

**Comment by nikic** [comment]
The core problem here is that `g()` can clobber `*p`, so we can only CSE the next `*p` load after the preceding `g()` call has been eliminated. Folding this reliably would require a single transform to do all of a) memory CSE b) comparison simplification based on assumes and c) control flow simplification to eliminate dead code.

This optimizes fine with a restrict pointer.

**Comment by zmodem** [comment]
I'm curious why the signedness makes a difference here though. Since we can fold it in the signed case, it seems we should be able to fold it in the unsigned too.

**Comment by nikic** [comment]
Ah yes. Looking a bit closer, GVN PREs the loads into the `g()` block, and past that point it should be possible to fold this based on only SSA analysis. This is quite elegant, I missed this option before.

For the signed case (with `-fwrapv` for a fair comparison) this then happens in the next InstCombine run, for the unsigned case it doesn't. At a guess, I think the reason is basically that we can materialize `nsw` on the add and then combine it into the `icmp`, but for the unsigned case we can't (the usual problem that sub canonicalized to add cannot represent unsigned nowrap flags). We could special-case this during icmp, or we could introduce additional nowrap flags to address the core issue.

**Comment by nikic** [comment]
Just to throw out another option, this could also be handled by CVP if it were to make use of conditions that become constants and then discards the code in the relevant branches.

If we ran SCCP with PredicateInfo that would probably also fix it.

**Comment by zmodem** [comment]
Right, I don't know what passes consume `llvm.assume`, but intuitively this seems similar to constant propagation. If one replaces `__builtin_assume(*p > 100)` with `*p = 100`, it optimizes very nicely.
